File: distinct_subsequences.js
// Distinct Subsequences: Count unique subsequences

var numDistinct = function(s, t) {
    let m = s.length, n = t.length;
    let dp = Array(n + 1).fill(0);
    dp[0] = 1;
    
    for (let i = 1; i <= m; i++) {
        for (let j = n; j >= 1; j--) {
            if (s[i-1] === t[j-1]) {
                dp[j] += dp[j-1];
            }
        }
    }
    
    return dp[n];
};

console.log(numDistinct("rabbbit", "rabbit"));
console.log(numDistinct("babgbag", "bag"));

File: distinct_subsequences.rb
# Distinct Subsequences: Count number of distinct subsequences of s matching t
def num_distinct(s, t)
  m, n = s.size, t.size
  dp = Array.new(m + 1) { Array.new(n + 1, 0) }
  (0..m).each { |i| dp[i][0] = 1 }
  (1..m).each do |i|
    (1..n).each do |j|
      dp[i][j] = dp[i - 1][j]
      dp[i][j] += dp[i - 1][j - 1] if s[i - 1] == t[j - 1]
    end
  end
  dp[m][n]
end

puts num_distinct("rabbbit", "rabbit")
puts num_distinct("babgbag", "bag")
